### ISSUE TITLE
[1.0] Add unit tests

### ISSUE THREAD

**Issue by micjamking** [micjamking/devtools-author#19]
Create unit tests for setting up the extension dependencies and communication with `chrome.storage`. A majority of the functionality is reliant on the successful return of several XHR's and the unit tests should reflect this behavior. Third-party functionality (ie. `chrome.storage` and `chrome.devtools.panels`) should be stubbed out appropriately.

## Unit testing scope

**DevTools Extension: [app/scripts/devtools.js](https://github.com/micjamking/devtools-author/blob/master/app/scripts/devtools.js)**
- **app.loadTheme:** gets stylesheet name from `chrome.storage` via XHR and fires callback function if returns success 200
- **app.init:** creates 'Author Settings' panel via `chrome.devtools.panels` and loads panel html page

**Author Settings panel: [app/scripts/panel.js](https://github.com/micjamking/devtools-author/blob/master/app/scripts/panel.js)**
- **init:** gets `theme.json` via XHR and fires callback function if returns success 200
- **getTheme:** gets `devtools-theme` from `chrome.storage` and returns it for display in UI
- **getFontFamily:** gets `devtools-fontFamily` from `chrome.storage` and returns it for display in UI
- **getFontSize:** gets `devtools-fontSize` from `chrome.storage` and returns it for display in UI

**References:**
- http://stackoverflow.com/questions/2869827/how-to-test-chrome-extensions
- http://stackoverflow.com/questions/14798528/testing-browser-extensions/17370531#17370531
- https://github.com/vitalets/sinon-chrome

**Comment by benniemosher** [comment]
@micjamking Has this been started at all anywhere? What would you like the approach to be? How can I help with this?

**Comment by micjamking** [comment]
Hey @benniemosher, thanks for asking! I'm new to testing so I could really use help here anywhere you're able to provide it. I started a test branch about 6 months ago using `sinon-chrome`, but it's really behind and I'm not sure if it's worth salvaging. However, it is there for reference: https://github.com/micjamking/devtools-author/tree/tests

I've updated the original issue description to include more details surrounding the testing scope and individual unit test expectations: https://github.com/micjamking/devtools-author/issues/19#issue-106215871

From what I remember in my last attempt, [`sinon-chrome`](https://github.com/acvetkov/sinon-chrome) was the recommended approach for mocking the `chrome.*` API.  But outside of that, I'm not totally sure the best way to approach it. I'm familiar with writing Jasmine testing, so if you use that or something similar, that would be cool. Very happy have you helping out @benniemosher!

:v:

**Comment by benniemosher** [comment]
Great man! After I get my personal site updated and launched I will get on this as the next task in my free time!

**Comment by micjamking** [comment]
@benniemosher Awesome! Again, I really appreciate the help, anywhere and anytime you can. 

Also, I'm interested in checking out your personal site when it's done :smile:

**Comment by benniemosher** [comment]
@micjamking I created a PR for setting up the testing framework. Have you had a chance to look at it yet?

**Comment by micjamking** [comment]
@benniemosher Sorry man, haven't had the chance to look at it yet. I plan to check it out later today and will merge in. My apologies for keeping you waiting.

**Comment by thgaskell** [comment]
Do you guys need help with this?

**Comment by micjamking** [comment]
Hey @thgaskell thanks for asking! 

re: PR #120; @benniemosher was able to get the initial TDD framework setup using [karma](http://karma-runner.github.io/0.13/index.html) + [jasmine](http://jasmine.github.io/) + [sinon-chrome](https://github.com/acvetkov/sinon-chrome) (via the [_karma-sinon-chrome_](https://github.com/9joneg/karma-sinon-chrome) plugin), however we're currently unable to test the `chrome.devtools.*` API because **_karma-sinon-chrome_ has a dependency on _sinon-chrome_ v0.2.1** and the `chrome.devtools.*` API was not added in _sinon-chrome_ until [v1.0.2](https://github.com/acvetkov/sinon-chrome/releases/tag/1.0.2). 

I merged in the PR so we could continue progress with setting up tests for the other parts not dependent on that specific API, but I think we need to fork _karma-sinon-chrome_ and get it working with the latest version _sinon-chrome_ so we can complete the full testing suite. Do you think this is something you could help out with?

Again, appreciate help anywhere anyone can provide it!

:v:

**Comment by thgaskell** [comment]
@micjamking definitely.

Thanks for the update. I'll look into getting the tests to work with the newer version of sinon-chrome.

:white_check_mark:

**Comment by thgaskell** [comment]
I have a forked version of karma-sinon-chrome and opened a PR (9joneg/karma-sinon-chrome#2) in hopes of getting that to be updated.

In the mean time, do you want to just link to the forked repo?
https://github.com/thgaskell/karma-sinon-chrome

**Comment by benniemosher** [comment]
My vote is that we do that in an isolated branch until it gets updated in
their master. It will at least allow us to get testing.



--

Thanks,

Bennie Mosher
https://benniemosher.com
+1 (970) 590-2040

**Comment by micjamking** [comment]
@benniemosher I'm fine with that; I'll  branch off current `master` and make a new `test` branch (the current one is 6 months behind). 

I'll update the package.json dependency to point to @thgaskell's fork in the new `test` branch.

```json
"devDependencies": {
    "karma-sinon-chrome": "thgaskell/karma-sinon-chrome#sinon-chrome-v1.1"
  }
```
Thanks again for helping out guys! :v:

**Comment by micjamking** [comment]
@benniemosher & @thgaskell the new testing branch is available here: [micjamking/devtools-author#tests](https://github.com/micjamking/devtools-author/tree/tests)

I've also added basic tests for the **`chrome.*`** API's we need to confirm they are working: c4377fd1692f0ac04682678fbb9c49a553a78996

I'll get started creating the initial tests based on the scope I've outline in the original comment: https://github.com/micjamking/devtools-author/issues/19#issue-106215871

This is extremely helpful guys! I owe you both :beers:

**Comment by micjamking** [comment]
Also removed tests from `master` branch for now, until tests are complete in `test` branch. 8826281

**Comment by thgaskell** [comment]
@micjamking I started writing some tests, but it's impossible to directly test the source methods since they're wrapped in **IIFE**s.

Also, the `init` functions are automatically ran as soon as the source files are included into karma, so it's trickier to mock the ajax requests since they're already called before the tests have ran.

It would be easier to test if the methods were exposed so the tests could get at them. Though, it would mean an _additional_ build step to re-encapsulate the code for distribution.

:package: Modules?

**Comment by micjamking** [comment]
Hey @thgaskell & @benniemosher 

Sorry for the delayed reply; I've been a bit busy with work and teaching the last week. That said, I was able to get some coding time in this weekend and we officially have tests now!!! :tada: 

The latest version of the tests are available here:  [micjamking/devtools-author#tests](https://github.com/micjamking/devtools-author/tree/tests)

I have an open pull request to merge the `test` branch into `master` here: https://github.com/micjamking/devtools-author/pull/129

I will eventually move to ES6 modules, but since I plan on rewriting **`panel.js`** using Angular 2 to address https://github.com/micjamking/devtools-author/issues/114, I decided on a simpler refactoring for now using a [revealing module pattern](https://addyosmani.com/resources/essentialjsdesignpatterns/book/#revealingmodulepatternjavascript) exposing just the methods I want to test and exporting each module as a global by attaching it to `window`. I've also added `grunt-karma` so I can run the tests as part of my build process as well.

For now, if I could get ya'll's review on PR #129 for feedback since I'm kinda green to testing, that would be super helpful. Mahalo!

:v:

**Comment by micjamking** [comment]
Hey @benniemosher & @thgaskell!

I just wanted to say thanks again for your guy's contributions in helping me get tests setup for DevTools Author. I've merged in #129 and added a `.travis.yml` to run CI builds.

[![screen shot 2016-03-06 at 9 24 01 am](https://cloud.githubusercontent.com/assets/651702/13556433/bef490e6-e37d-11e5-806a-d810c017a365.png)](https://travis-ci.org/micjamking/devtools-author)

Much Mahalos!
:v:
